pkg/cache/sql/informer/listoption_indexer.go

```diff
@@ -0,0 +1,455 @@
+
+// afterUpsert saves sortable/filterable fields into tables
+func (l *ListOptionIndexer) afterUpsert(key string, obj any, tx db.TXClient) error {
+	args := []any{key}
+	for _, field := range l.indexedFields {
+		value, err := getField(obj, field)
+		if err != nil {
+			logrus.Errorf("cannot index object of type [%s] with key [%s] for indexer [%s]: %v", l.GetType().String(), key, l.GetName(), err)
+			cErr := tx.Cancel()
+			if cErr != nil {
+				return errors.Wrap(err, cErr.Error())
+			}
+			return err
+		}
+		switch typedValue := value.(type) {
+		case nil:
+			args = append(args, "")
+		case int, bool, string:
+			args = append(args, fmt.Sprint(typedValue))
+		case []string:
+			args = append(args, strings.Join(typedValue, "|"))
+		default:
+			return errors.Errorf("%v has a non-supported type value: %v", field, value)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	return tx.StmtExec(tx.Stmt(l.addField), args...)
+}
+
+func (l *ListOptionIndexer) afterDelete(key string, tx db.TXClient) error {
+	args := []any{key}
+	return tx.StmtExec(tx.Stmt(l.deleteField), args...)
+}
+
+// ListByOptions returns objects according to the specified list options and partitions
+// result is an unstructured.UnstructuredList, the continue token for the next page (or an error)
+func (l *ListOptionIndexer) ListByOptions(ctx context.Context, lo ListOptions, partitions []partition.Partition, namespace string) (*unstructured.UnstructuredList, string, error) {
+	// 1- Intro: SELECT and JOIN clauses
+	stmt := fmt.Sprintf(`SELECT o.object, o.objectnonce, o.dek, o.deknonce FROM "%s" o`, l.GetName())
+	stmt += "\n  "
+	stmt += fmt.Sprintf(`JOIN db2."%s_fields" f ON o.key = f.key`, l.GetName())
+	params := []any{}
+
+	// 2- Filtering: WHERE clauses (from lo.Filters)
+	whereClauses := []string{}
+	for _, orFilters := range lo.Filters {
+		orClause, orParams := buildORClause(orFilters)
+		whereClauses = append(whereClauses, orClause)
+		params = append(params, orParams...)
+	}
+
+	// WHERE clauses (from namespace)
+	if namespace != "" && namespace != "*" {
+		whereClauses = append(whereClauses, fmt.Sprintf(`f."metadata.namespace" = ?`))
+		params = append(params, namespace)
+	}
+
+	// WHERE clauses (from partitions and their corresponding parameters)
+	partitionClauses := []string{}
+	for _, partition := range partitions {
+		if partition.Passthrough {
+			// nothing to do, no extra filtering to apply by definition
+		} else {
+			singlePartitionClauses := []string{}
+
+			// filter by namespace
+			if partition.Namespace != "" && partition.Namespace != "*" {
+				singlePartitionClauses = append(singlePartitionClauses, fmt.Sprintf(`f."metadata.namespace" = ?`))
+				params = append(params, partition.Namespace)
+			}
+
+			// optionally filter by names
+			if !partition.All {
+				names := partition.Names
+
+				if len(names) == 0 {
+					// degenerate case, there will be no results
+					singlePartitionClauses = append(singlePartitionClauses, "FALSE")
```

Comment:
Nit: Rather than appending a clause which will return no results, could we instead just return some sort of default here? I'm open to other ideas though,either case is probably ok.

Reply:
We have three cases to "encode" into `singlePartitionClauses`:

1. the normal case: one or two clauses to be `AND`ed together as filtering (WHERE) conditions. Currently encoded as a non-empty list of clauses
2. the `partition.All` degenerate case: no filtering conditions. Currently encoded as an empty list of clauses
3. the no-Names degenerate case: no results _from this partition_. Currently encoded as `FALSE`

Note that 3. does not mean we can immediately `return toUnstructuredList([]any{}), nil, nil` - other partitions might have results and all `singlePartitionClauses` have to be `OR`ed together.

We could encode 3. in another variable (eg a bool) to avoid the `OR (FALSE)` "nop" clause - `singlePartitionNoResults` maybe? I was not sure this is actually easier to comprehend.